subprojects/execution/src/main/java/org/gradle/internal/execution/UnitOfWork.java

```diff
@@ -134,14 +138,136 @@ default void visitIdentityInputs(InputVisitor visitor) {}
      */
     default void visitRegularInputs(InputVisitor visitor) {}
 
+    interface InputVisitor {
+        default void visitInputProperty(
+            String propertyName,
+            ValueSupplier value
+        ) {}
+
+        default void visitInputFileProperty(
+            String propertyName,
+            InputPropertyType type,
+            InputFileValueSupplier value
+        ) {}
+
+    }
+
+    enum InputPropertyType {
+        /**
+         * Non-incremental inputs.
+         */
+        NON_INCREMENTAL(false, false),
+
+        /**
+         * Incremental inputs.
+         */
+        INCREMENTAL(true, false),
+
+        /**
+         * These are the primary inputs to the incremental work item;
+         * if they are empty the work item shouldn't be executed.
+         */
+        PRIMARY(true, true);
+
+        private final boolean incremental;
+        private final boolean skipWhenEmpty;
+
+        InputPropertyType(boolean incremental, boolean skipWhenEmpty) {
+            this.incremental = incremental;
+            this.skipWhenEmpty = skipWhenEmpty;
+        }
+
+        public boolean isIncremental() {
+            return incremental;
+        }
+
+        public boolean isSkipWhenEmpty() {
+            return skipWhenEmpty;
+        }
+    }
+
+    interface ValueSupplier {
+        @Nullable
+        Object getValue();
+    }
+
+    interface FileValueSupplier extends ValueSupplier {
```

Comment:
I wonder if we can come up with a different name for this. `FileValueSupplier` suggests that the supplied value is a file or file contents, but this is not the case, since the value can still be anything.

One suggestion I can think of is `ValueWithFilesSupplier`. This clearly states that files are an additional thing on top of the value which aligns with the added `getFiles` method. This further makes `InputValueWithFilesSupplier` and `OutputValueWithFilesSupplier` better readable. The `OutputValueWithFilesSupplier` name also makes a clear distinction between the output value, which is by itself a file, and the additional files that come with it when it's a directory.

Reply:
This is used to represent _things that will be resolved to zero or more files._ The actual object representing the files might be of a range of things like a `File`, a `Set<Path>`, a `RegularFileProperty`, a `Callable<File>` a `Callable<Provider<FileSystemLocation>>`, a `ConfigurableFileCollection` etc., that's why the type is an `Object` here.

I think what we could do to disambiguate this is to a) add Javadoc, and b) rename `ValueSupplier.getValue()` to `getUnresolvedValue()`. Lemme take a stab at this.